packages/contracts-sdk/src/implementations/farcaster/FarcasterKeyGatewayContract.ts

```diff
@@ -0,0 +1,248 @@
+  extends BaseContract<FarcasterKeyGatewayContractError>
+  implements IFarcasterKeyGatewayContract
+{
+  constructor(
+    protected providerOrSigner: ethers.Provider | ethers.Signer,
+    protected contractAddress: EVMContractAddress,
+  ) {
+    super(
+      providerOrSigner,
+      contractAddress,
+      ContractsAbis.FarcasterKeyGatewayAbi.abi,
+    );
+  }
+
+  public nonces(
+    address: EVMAccountAddress,
+  ): ResultAsync<
+    bigint,
+    FarcasterKeyGatewayContractError | BlockchainCommonErrors
+  > {
+    return ResultAsync.fromPromise(
+      this.contract.nonces(address) as Promise<bigint>,
+      (e) => {
+        return this.generateError(e, "Unable to call nonces()");
+      },
+    );
+  }
+
+  public add(
+    key: string,
+    encodedMetadata: string, // obtained from getEncodedSignedKeyRequestMetadata()
+    overrides?: ContractOverrides,
+  ): ResultAsync<
+    WrappedTransactionResponse,
+    FarcasterKeyGatewayContractError | BlockchainCommonErrors
+  > {
+    // Based on the docs, only keyType and metadataType of 1 is supported at the moment
+    // https://docs.farcaster.xyz/reference/contracts/reference/key-gateway#add
+    return this.writeToContract("add", [1, key, 1, encodedMetadata], overrides);
+  }
+
+  // Called by third party on behalf of an fid
+  // Called by the owner's EW to sign a 'Register' signature
+  // The signature and encodedMetadata for the 'Add' function must first be obtained using the wallet owning the fid via getSignedKeyRequestSignatureAndEncodedMetadata()
+  public addFor(
+    fidOwnerAddress: EVMAccountAddress,
+    keyToAdd: string,
+    encodedMetadata: string,
+    deadline: UnixTimestamp,
+    signature: SignedAddSignature,
+    overrides?: ContractOverrides,
+  ): ResultAsync<
+    WrappedTransactionResponse,
+    FarcasterKeyGatewayContractError | BlockchainCommonErrors
+  > {
+    // Based on the docs, only keyType and metadataType of 1 are supported at the moment
+    // https://docs.farcaster.xyz/reference/contracts/reference/key-gateway#addfor
+    const keyGatewayAddForParams = {
+      keyType: 1, // 1 is the only key type for now
+      key: keyToAdd, // the key to be registered - doesn't have to be same key as mainUser, could be a newly generated key
+      metadataType: 1,
+      metadata: encodedMetadata,
+      sig: signature,
+      deadline: deadline,
+    };
+
+    return this.writeToContract(
+      "addFor",
+      [
+        fidOwnerAddress,
+        keyGatewayAddForParams.keyType,
+        keyGatewayAddForParams.key,
+        keyGatewayAddForParams.metadataType,
+        keyGatewayAddForParams.metadata,
+        keyGatewayAddForParams.sig,
+        keyGatewayAddForParams.deadline,
+      ],
+      overrides,
+    );
+  }
+
+  public getAddSignature(
+    ownerAddress: EVMAccountAddress,
+    keyToAdd: EVMAccountAddress,
+    encodedMetadata: EncodedSignedKeyRequestMetadata,
+    deadline: UnixTimestamp,
+  ): ResultAsync<
+    SignedAddSignature,
```

Comment:
Similar to the SignedRegisterSignature, consider FarcasterAddSignature as the type name